### PR TITLE
Add api key validation

### PR DESCRIPTION
**Pull request by davidbalazs93**
As part of productionize Notifications Push work, we are introducing validation for apiKeys via Mashery call.
Please see https://sites.google.com/a/ft.com/universal-publishing/documentation/story-specs/productionising-notifications-push-api for architecture diagram and specifications.

Changes in this PR:
 - added vendoring
 - added HTTP call to Mashery to validate apiKey
 - updated Dockerfile

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -144,10 +145,10 @@ func main() {
 		}
 
 		queueHandler := consumer.NewMessageQueueHandler(whitelistR, mapper, dispatcher)
-		httpClient := &http.Client{}
-		consumer := queueConsumer.NewBatchedConsumer(consumerConfig, queueHandler.HandleMessage, httpClient)
+		hc := getResilientClient()
```
Comment (davidbalazs93) [inline app.go]: lgtm, only a small obs. wouldn't it be more readable to have httpClient instead of hc?

Comment (tosan88) [inline app.go]: Sure, I would have preferred an interface, but later on decided to not use as it would have been a bigger change

Comment (davidbalazs93) [inline app.go]: :+1:

File: resources/push.go [hunk 2]
```diff
@@ -8,17 +8,39 @@ import (
 
 	"github.com/Financial-Times/notifications-push/dispatcher"
 	log "github.com/Sirupsen/logrus"
+	"io"
+	"io/ioutil"
 )
 
+const (
+	ApiKeyHeaderField = "X-Api-Key"
```
Comment (danielcopaciu) [inline resources/push.go]: Why are these constants public?

File: resources/push.go [hunk 3]
```diff
@@ -73,3 +95,50 @@ func getClientAddr(r *http.Request) string {
...
+	log.WithField("url", req.URL.String()).WithField("apiKeyFirstChars", apiKeyFirstChars).Info("Calling Mashery to validate api key")
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		log.WithField("url", req.URL.String()).WithError(err).Error("Cannot send request to Mashery")
+		http.Error(w, "Request to validate api key failed", http.StatusInternalServerError)
+		return false
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, resp.Body)
```
Comment (danielcopaciu) [inline resources/push.go]: Why is this necessary?

File: resources/push.go [hunk 4]
```diff
@@ -73,3 +95,50 @@ func getClientAddr(r *http.Request) string {
...
+		http.Error(w, "Invalid api key", http.StatusUnauthorized)
+		return false
+	}
+
+	log.WithField("url", req.URL.String()).Errorf("Received unexpected status code from Mashery: %d", respStatusCode)
+	http.Error(w, "Request to validate api key returned an unexpected response", http.StatusServiceUnavailable)
+	return false
+}
+
+func isApiKeyFirstFourCharsLoggable(providedApiKey string) bool {
```
Comment (danielcopaciu) [inline resources/push.go]: Do we really need a separate function for this check?

File: app.go [hunk 5]
```diff
@@ -172,3 +179,18 @@ func server(listen string, resource string, dispatcher dispatcher.Dispatcher, hi
 	err := http.ListenAndServe(listen, nil)
 	log.Fatal(err)
 }
+
+func getResilientClient() *http.Client {
+	tr := &http.Transport{
```
Comment (tosan88) [inline app.go]: Indeed, it was a mistake (first it was _actually_ resilient, but later on decided to not use as it would have been a bigger change but didn't rename the method). I'll fix this as well

File: resources/push.go (hunk 4 above)
Comment (tosan88) [inline resources/push.go]: I wanted to be clear what the check is actually about. I'm happy to change that but I'd prefer this way rather then commenting on the spot

File: resources/push.go (hunk 3 above)
Comment (tosan88) [inline resources/push.go]: AFAIK this is needed for reusing HTTP connections, see http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body

File: resources/push.go [hunk 6]
```diff
@@ -73,3 +95,50 @@ func getClientAddr(r *http.Request) string {
...
+
+func validApiKey(w http.ResponseWriter, providedApiKey string, masheryApiKeyValidationURL string, httpClient *http.Client) bool {
+	req, err := http.NewRequest("GET", masheryApiKeyValidationURL, nil)
+	if err != nil {
+		log.WithField("url", masheryApiKeyValidationURL).WithError(err).Error("Invalid URL for api key validation")
+		http.Error(w, "Invalid URL", http.StatusInternalServerError)
+		return false
+	}
+
+	req.Header.Set(ApiKeyHeaderField, providedApiKey)
```
Comment (tosan88) [inline resources/push.go]: True, will add a check

File: resources/push.go (hunk 2 above)
Comment (tosan88) [inline resources/push.go]: No reason, will change that

File: resources/push.go [hunk 7]
```diff
@@ -73,3 +95,50 @@ func getClientAddr(r *http.Request) string {
 	}
 	return r.RemoteAddr
 }
+
+func validApiKey(w http.ResponseWriter, providedApiKey string, masheryApiKeyValidationURL string, httpClient *http.Client) bool {
```
Comment (kovacshuni) [inline resources/push.go]: `isValidApiKey`

Comment (kovacshuni) [inline resources/push.go]: could you move this function into a separate file, for the validation to be abstracted away, make it a separate responsibility entity.

File: resources/push.go (hunk 2 above)
Comment (tosan88) [inline resources/push.go]: Done

File: resources/push.go (hunk 6 above)
Comment (tosan88) [inline resources/push.go]: Done

File: resources/push.go (hunk 4 above)
Comment (tosan88) [inline resources/push.go]: Changed

File: app.go (hunk 5 above)
Comment (tosan88) [inline app.go]: Fixed

File: resources/push.go (hunk 7 above)
Comment (tosan88) [inline resources/push.go]: Fair enough. Changed

**Comment by kovacshuni** [comment]
i would like a separate entity for the validator like

`func (v validator) validate() {}`

so it can be inserted, injected, mocked, like in java. If you'd want to test the function that is using the validation function you'd have to set up everything for them. Otherwise you could just mock this component and mask away the validator's functionality /and the other way around.

Maybe in the next release.
